src/topology/stone_cech.lean

```diff
@@ -220,11 +220,9 @@ def stone_cech_unit (x : α) : stone_cech α := ⟦pure x⟧
 
 /-- The image of stone_cech_unit is dense. (But stone_cech_unit need
   not be an embedding, for example if α is not Hausdorff.) -/
-lemma stone_cech_unit_dense : closure (range (@stone_cech_unit α _)) = univ :=
-begin
-  convert quotient_dense_of_dense (eq_univ_iff_forall.mp dense_inducing_pure.closure_range),
-  rw [←range_comp], refl
-end
+lemma dense_range_stone_cech_unit : dense_range (@stone_cech_unit α _) :=
+-- With term mode proof Lean tries to use wrong instances
```

Comment:
This is a bit worrying. Do you understand what happens exactly?

Reply:
Here is the error message. I don't understand what happens here:
```
223:7: synthesized type class instance is not definitionally equal to expression inferred by typing rules, synthesized
  _inst_1
inferred
  ⊥
```

Reply:
Thank your for fixing this!

Reply:
Ok, I had a look. The core issue is our type-class system doesn't like arguments involving several topological structures on the same type. Given a bare type A (no topological structure), on can build `ultra_filter A`, a map `pure : A -> ultra_filter A` and a topology on `ultra_filter A`. The map `pure` has dense range. This fact doesn't need a topology on `A`. The next lemma was this map is dense inducing if `A` is endowed with the discrete topology.

Trouble comes with the Stone-Cech construction where `A` does have a topology, not necessarily discrete, but we still want to use that `pure` has a dense range. There type class inference can get confused easily, and this is what the error message is saying: `A` has both a completely unknown topology `_inst_1` and the discrete topology `⊥`.

I pushed two commits to this branch. The first one extracts the density statement that is independent of a topology on `A` and was buried in the proof of the dense inducing thing. The second one also cleans up the later statement by first replacing "pure is inducing if A is endowed with the discrete topology" by the less contrived "pure induces on A the discrete topology". We still need the contrived version because it gives access to the `dense_inducing` API and its crucial extension construction.